cfme/cloud/instance/__init__.py

```diff
@@ -306,54 +303,197 @@ def power_control_from_cfme(self, *args, **kwargs):
+        if provisioning.get('availability_zone'):
+            av_zone_entities = self.appliance.rest_api.collections.availability_zones.find_by(
+                name=provisioning['availability_zone'])
+            if av_zone_entities and av_zone_entities[0].ems_id == flavor.ems_id:
+                availability_zone_id = av_zone_entities[0].id
+        if not availability_zone_id and 'availability_zone_id' in cloud_subnet:
+            availability_zone_id = cloud_subnet['availability_zone_id']
+        if not availability_zone_id:
+            availability_zone_id, _ = wait_for(
+                _find_availability_zone_id, num_sec=100, delay=5,
+                message="availability_zone present")
+
+        # find out cloud tenant
+        cloud_tenant_id = None
+        tenant_name = provisioning.get('cloud_tenant')
+        if tenant_name:
+            for tenant in provider_data.get('cloud_tenants', []):
+                if (tenant['name'] == tenant_name and
+                        tenant['enabled'] and
+                        tenant['ems_id'] == flavor.ems_id):
+                    cloud_tenant_id = tenant['id']
+
+        resource_group_id = None
+        if self.provider.one_of(AzureProvider):
+            resource_groups = self.appliance.rest_api.get(
+                '{}?attributes=resource_groups'.format(provider_rest._href))['resource_groups']
+            resource_group_id = None
+            resource_group_name = provisioning.get('resource_group')
+            for res_group in resource_groups:
+                if (res_group['name'] == resource_group_name and
+                        res_group['ems_id'] == provider_rest.id):
+                    resource_group_id = res_group['id']
+                    break
+
+        inst_args = {
+            "version": "1.1",
+            "template_fields": {
+                "guid": image_guid
+            },
+            "vm_fields": {
+                "placement_auto": False,
+                "vm_name": self.name,
+                "instance_type": flavor_id,
+                "request_type": "template",
+                "cloud_network": cloud_network['id'],
+                "cloud_subnet": cloud_subnet['id'],
+                "placement_availability_zone": availability_zone_id,
+
+            },
+            "requester": {
+                "user_name": "admin",
+                "owner_email": "admin@example.com",
+                "auto_approve": True,
+            },
+            "tags": {
+            },
+            "ems_custom_attributes": {
+            },
+            "miq_custom_attributes": {
+            }
+        }
+        if cloud_tenant_id:
+            inst_args['vm_fields']['cloud_tenant'] = cloud_tenant_id
+
+        if resource_group_id:
+            inst_args['vm_fields']['resource_group'] = resource_group_id
+
+        if self.provider.one_of(EC2Provider):
+            inst_args['vm_fields']['monitoring'] = 'basic'
+
+        return inst_args
```

Comment:
**REQUIRED**
This method is too big. Please split it on several methods.

Reply:
simplified rest calls to use filters. but moving code to other helper methods won't be much useful as those need prerequisites and won't be used widely